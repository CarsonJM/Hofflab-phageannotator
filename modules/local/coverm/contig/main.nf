process COVERM_CONTIG {
    tag "$meta.id"
    label 'process_high'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/coverm:0.7.0--h07ea13f_1':
        'biocontainers/coverm:0.7.0--h07ea13f_1' }"

    input:
    tuple val(meta), path(fastq)
    tuple val(meta), path(fasta)

    output:
    tuple val(meta), path("${prefix}_alignment_results.tsv")    , emit: alignment_results
    tuple val(meta), path("${prefix}.bam")                      , emit: bam
    path "versions.yml"                                         , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    prefix = task.ext.prefix ?: "${meta.id}"
    """
    if [ \$(zcat ${fastq[0]} | grep "@" | wc -l) -gt 0 ]; then
        coverm \\
            contig \\
            -1 ${fastq[0]} \\
            -2 ${fastq[1]} \\
            --reference ${fasta} \\
            --output-file ${prefix}_alignment_results.tsv \\
            --bam-file-cache-directory ${prefix}_bam_files/ \\
            --threads ${task.cpus} \\
            ${args}

        mv ${prefix}_bam_files/*.bam ${prefix}.bam
    else
        touch ${prefix}_alignment_results.tsv
        touch ${prefix}.bam
    fi

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        coverm: \$(echo \$(coverm --version 2>&1) | sed 's/^.*coverm //; s/Using.*\$//' ))
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    mkdir -p ${prefix}_bam_files
    touch ${prefix}_alignment_results.tsv
    touch ${prefix}.bam

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        coverm: \$(echo \$(coverm --version 2>&1) | sed 's/^.*coverm //; s/Using.*\$//' ))
    END_VERSIONS
    """
}
