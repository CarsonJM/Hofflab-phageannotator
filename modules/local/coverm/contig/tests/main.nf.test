nextflow_process {

    name "Test Process: COVERM_CONTIG"
    script "../main.nf"
    process "COVERM_CONTIG"


    test("fastq.gz single + fasta.gz") {
        when {
            process {
                """
                input[0] = [
                    [ id: 'test1', single_end:true ],
                    [
                        file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz', checkIfExists: true),
                    ]
                ]
                input[1] = [
                    [ id: 'test1' ],
                    file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/fasta/test1.contigs.fa.gz', checkIfExists: true)
                ]
                """
            }
        }
        then {
            assertAll (
                { assert process.success },
                { assert snapshot(
                    process.out.alignment_results,
                    file(process.out.bam[0][1]).name,
                    file(process.out.log[0][1]).name,
                    process.out.versions
                    ).match()
                }
            )
        }
    }


    test("fastq.gz paired + fasta.gz") {
        when {
            process {
                """
                input[0] = [
                    [ id: 'test1', single_end:false ],
                    [
                        file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz', checkIfExists: true)
                    ]
                ]
                input[1] = [
                    [ id: 'test1' ],
                    file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/fasta/test1.contigs.fa.gz', checkIfExists: true)
                ]
                """
            }
        }
        then {
            assertAll (
                { assert process.success },
                { assert snapshot(
                    process.out.alignment_results,
                    file(process.out.bam[0][1]).name,
                    file(process.out.log[0][1]).name,
                    process.out.versions
                    ).match()
                }
            )
        }
    }


    test("fastq.gz + fasta.gz - stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = [
                    [ id: 'test1' ],
                    [
                        file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz', checkIfExists: true)
                    ]
                ]
                input[1] = [
                    [ id: 'test1' ],
                    file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/fasta/test1.contigs.fa.gz', checkIfExists: true)
                ]
                """
            }
        }
        then {
            assertAll (
                { assert process.success },
                { assert snapshot(process.out).match() },
            )
        }
    }
}
