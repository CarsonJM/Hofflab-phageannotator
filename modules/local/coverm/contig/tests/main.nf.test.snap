{
    "fastq.gz single + fasta.gz": {
        "content": [
            [
                [
                    {
                        "id": "test1",
                        "single_end": true
                    },
                    "test1_alignment_results.tsv:md5,a23785959ca535474b095a8511293dc3"
                ]
            ],
            "test1.bam",
            "test1.log",
            [
                "versions.yml:md5,2bede4e7376363a3d53626ecbcac2b8c"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.1"
        },
        "timestamp": "2024-07-24T10:11:47.823239765"
    },
    "fastq.gz + fasta.gz - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test1"
                        },
                        "test1_alignment_results.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test1"
                        },
                        "test1.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test1"
                        },
                        "test1.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    "versions.yml:md5,2bede4e7376363a3d53626ecbcac2b8c"
                ],
                "alignment_results": [
                    [
                        {
                            "id": "test1"
                        },
                        "test1_alignment_results.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "bam": [
                    [
                        {
                            "id": "test1"
                        },
                        "test1.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "log": [
                    [
                        {
                            "id": "test1"
                        },
                        "test1.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,2bede4e7376363a3d53626ecbcac2b8c"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.1"
        },
        "timestamp": "2024-07-24T10:12:32.050774863"
    },
    "fastq.gz paired + fasta.gz": {
        "content": [
            [
                [
                    {
                        "id": "test1",
                        "single_end": false
                    },
                    "test1_alignment_results.tsv:md5,7a91af5bae97e185cb910024a0bad0ac"
                ]
            ],
            "test1.bam",
            "test1.log",
            [
                "versions.yml:md5,2bede4e7376363a3d53626ecbcac2b8c"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.1"
        },
        "timestamp": "2024-07-24T10:12:10.764688564"
    }
}